### ISSUE TITLE
Italic/bold is treated as a list, When Italic/bold is written on the next line of the list

### ISSUE THREAD

**Issue by yama-syo** [markedjs/marked#1980]
**Marked version:**
  
master

**Describe the bug**
  
To be exact, this problem is
After recognizing the list, determine whether the next line is the next list only by the symbol(without space) of the list.
  
In detail  
The next line after the line recognized as a list  
The next line is the symbol(\* ,+ ,- ,1.) used in the list followed by a character without space.  
In the above case, marked determines that the next line is also a list.  

Therefore, "\*" is used in the list symbol, so if there is an "\*" in the next line, it is judged as a list.
  
  
**To Reproduce**  

[editor](https://marked.js.org/demo/?text=*%20list1%0A*Italic%2Fnot%20list*%0A*%20list2%0A%0A%0A*%20list1%0A**bold%2Fnot%20list**%0A*%20list2%0A%0A%0A1.%20list1%0A1.not%20list%0A1.%20list2%0A%0A%0A%0A1.%20%5Bx%5D%20list1%0A1.%5Bx%5D%20not%20list%0A1.%20%5Bx%5D%20list2&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)

**Expected behavior**
  
I applied the following self-made patch to this problem.

```
diff --git a/src/rules.js b/src/rules.js
index d3175e8..a83780d 100644
--- a/src/rules.js
+++ b/src/rules.js
@@ -43,7 +43,7 @@ block.def = edit(block.def)
   .getRegex();
 
 block.bullet = /(?:[*+-]|\d{1,9}[.)])/;
-block.item = /^( *)(bull) ?[^\n]*(?:\n(?! *bull ?)[^\n]*)*/;
+block.item = /^( *)(bull) +[^\n]*(?:\n(?! *bull +)[^\n]*)*/;
 block.item = edit(block.item, 'gm')
   .replace(/bull/g, block.bullet)
   .getRegex();
```

**note**
  
This sentence is through google translate.
Did you get the content?

**Comment by UziTech** [comment]
You "editor" link doesn't have a `*` in it. I am confused by what the problem markdown is.

Can you provide the markdown that has a problem?

**Comment by yama-syo** [comment]
Sorry,


**To Reproduce**  


https://marked.js.org/demo/?outputType=lexer&text=%23%20header%20%0A%20%20%0Aparagraph%0A&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master

**Comment by UziTech** [comment]
Thanks for the reproduction.

[marked demo](https://marked.js.org/demo/?text=*%20list1%0A*Italic%2Fnot%20list*%0A*%20list2%0A%0A%0A*%20list1%0A**bold%2Fnot%20list**%0A*%20list2%0A%0A%0A1.%20list1%0A1.not%20list%0A1.%20list2%0A%0A%0A%0A1.%20%5Bx%5D%20list1%0A1.%5Bx%5D%20not%20list%0A1.%20%5Bx%5D%20list2&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)
[commonmark demo](https://spec.commonmark.org/dingus/?text=*%20list1%0A*Italic%2Fnot%20list*%0A*%20list2%0A%0A%0A*%20list1%0A**bold%2Fnot%20list**%0A*%20list2%0A%0A%0A1.%20list1%0A1.not%20list%0A1.%20list2%0A%0A%0A%0A1.%20%5Bx%5D%20list1%0A1.%5Bx%5D%20not%20list%0A1.%20%5Bx%5D%20list2)

It looks like this may be unspecified behavior. Are you looking for something closer to the commonmark demo where the bold line is part of the previous list item or is the bold not supposed to be in the list at all?

**Comment by yama-syo** [comment]
I think this issue is a list-only issue.  
Because, as I wrote in In detail, marked doesn't care about the space needed in the list.  
This is the main problem.  


As a test, compare A and B with list problems (without Italic / bold)  

[marked demo](https://marked.js.org/demo/?text=%23%23%20%2Blist%0A%0A%2B%20list1%0A%2BNot%20list(without%20space)%0A-%20list2%0A%0A%23%23%20-list%0A%0A-%20list1%0A-Not%20list(without%20space)%0A-%20list2%0A%0A%23%23%20number(1.)list%20%0A1.%20list%0A1.Notlist(without%20space)%0A1.%20list%0A&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)  

[commonmark demo](https://spec.commonmark.org/dingus/?text=%23%23%20%2Blist%0A%0A%2B%20list1%0A%2BNot%20list(without%20space)%0A-%20list2%0A%0A%23%23%20-list%0A%0A-%20list1%0A-Not%20list(without%20space)%0A-%20list2%0A%0A%23%23%20number(1.)list%20%0A1.%20list%0A1.Notlist(without%20space)%0A1.%20list%0A)

For commonmark, each list has 2 lists.  
For markd, each list has 3 lists.  
The list generation is different.

**Comment by UziTech** [comment]
If you would like to create a PR we could get your patch merged.

**Comment by calculuschild** [comment]
Looks like this is *mostly* solved by #2112. There is still a difference of Marked creating `<p>` tags around the "not-list" items and Commomark leaving them as inline text, but at least Italics / Bold are no longer treated as list items.

**Comment by jangxyz** [comment]
I have came upon this issue as well, having problems like the [demo](https://marked.js.org/demo/?text=%23%23%20%2Blist%0A%0A%2B%20list1%0A%2BNot%20list(without%20space)%0A-%20list2%0A%0A%23%23%20-list%0A%0A-%20list1%0A-Not%20list(without%20space)%0A-%20list2%0A%0A%23%23%20number(1.)list%20%0A1.%20list%0A1.Notlist(without%20space)%0A1.%20list%0A&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master) above.

**TLDR;**
1. Why marked is not parsing unspaced bullets correclty.
2. What others are involved in the problem.
3. I have a fix.



### What is wrong

Here's what I have understood through reading the `list` block parser in Tokenzier ([Tokenzier.js#165](https://github.com/markedjs/marked/blob/4c5b974b391f913ac923610bd3740ef27ccdae95/src/Tokenizer.js#L165)).
1. while looping through lines,
4. checks whether this line looks like an item with `itemRegex` pattern.
5. check the next line if it looks like an item with `nextBulletRegex`. (Another while loop)
    If it is, that's it for this item. If it isn't, keep adding it up to the current item.

I've found that while `itemRegex` testing for an item bullet is okay, `nextBulletRegex` is incomplete, thus mistakenly stopping when meeting a bullet without a space.

``` javascript
var itemRegex = new RegExp(`^( {0,3}${bull})((?: [^\\n]*)?(?:\\n|$))`);
// |^( {0,3}${bull})((?: [^\\n]*)?(?:\\n|$))|
//   1. |^|          : start of sentence
//   2. | {0,3}|     : 0~3 spaces available
//   3. |${bull}|    : pre-computed bullet character
//   4. |( [^\\n]*)?|: a space, optionally followed by non-newline characters (optional)
//   5. |(\\n|$)|    : end of sentence, or new line

const nextBulletRegex = new RegExp(`^ {0,${Math.min(3, indent - 1)}}(?:[*+-]|\\d{1,9}[.)])`);
//  |^ {0,${Math.min(3, indent - 1)}}(?:[*+-]|\\d{1,9}[.)])|
//    1. |^ |                            : start of sentence, followed by a space
//    2. |{0,${Math.min(3, indent - 1)}}|: zero to some predefined indent number of spaces (string evaluation)
//    3. |([*+-]|\\d{1,9}[.)])|          : any item bullet, either ordered or unordered
//
//    NOTE does not check whether there is any whitespace after the bullet.
```

Say we have a smaller sample ([demo](https://marked.js.org/demo/?outputType=html&text=-%20item1%0A-%20item2%0A-Not%20a%20list%20item(without%20space)%0A-%20item3%0A&options=%7B%0A%20%22baseUrl%22%3A%20null%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22headerIds%22%3A%20true%2C%0A%20%22headerPrefix%22%3A%20%22%22%2C%0A%20%22highlight%22%3A%20null%2C%0A%20%22langPrefix%22%3A%20%22language-%22%2C%0A%20%22mangle%22%3A%20true%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22sanitize%22%3A%20false%2C%0A%20%22sanitizer%22%3A%20null%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22smartLists%22%3A%20false%2C%0A%20%22smartypants%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%2C%0A%20%22xhtml%22%3A%20false%0A%7D&version=master)):

```markdown
- item1
- item2
-Not a list item(without space)
- item3
```

The current implementation of `nextBulletRegex` does not check whether there is a whitespace after the bullet, and breaks out the loop as it meets `"-Not a list item(without space)"`.

https://github.com/markedjs/marked/blob/4c5b974b391f913ac923610bd3740ef27ccdae95/src/Tokenizer.js#L241-L243

On next iteration it figures that `"-Not a list item(without space)"` is just a text and ends the list, and starts a new paragraph.

### The fix (or, fixes)

I tried changing `nextBulletRegex` to check for an additional whitespace like `itemRegex` does.

```javascript
//const nextBulletRegex = new RegExp(`^ {0,${Math.min(3, indent - 1)}}(?:[*+-]|\\d{1,9}[.)])`);
const nextBulletRegex = new RegExp(`^ {0,${Math.min(3, indent - 1)}}(?:[*+-]|\\d{1,9}[.)])((?: [^\\n]*)?(?:\\n|$))`);
```

Indeed **this will render the above sample correctly**,
but unfortunately broke other tests, colliding with

<details>
<summary>
thematic breaks (---): 
</summary>

Commonmark [Example#57](https://spec.commonmark.org/0.30/#example-56)
```markdown
- foo
***
- bar
```
 
[Example#99](https://spec.commonmark.org/0.30/#example-99)
```markdown
- foo
-----
```
</details>

<details>
<summary>
setext headings (---):
</summary>

Commonmark [Example#300](https://spec.commonmark.org/0.30/#example-300)
```markdown
- # Foo
- Bar
  ---
  baz
```
</details>

These all have the form of 'bullet-like' block syntax inside the list, mingled with various indent options and precedences over each other (setext heading has higher order than thematic breaks).

*Maybe* the reason `nextBulletRegex` did not have any whitespace in the first place was to make it compatible with these other syntaxes. However, it does not solve the problem in this issue, and at minimum is a misnomer. I think it should be fixed.

### Additional Touch

I did manage to make all tests pass, with adding additional logic inside the second loop.

```javascript
while (src) {
  rawLine = src.split('\n', 1)[0];
  prevLine = line;
  line = rawLine;

  // ...

  if (this.rules.block.hr.test(line)) {
    // make sure it is not a setext heading, which takes precedence
    const twoLines = [prevLine, line].join('\n');
    const matchLheadingPattern = this.rules.block.lheading.test(twoLines);
    const lineIdentIsLargerThanPrev = line.search(/[^ ]/) >= indent;
    const isNextLineSetextHeading = matchLheadingPattern && lineIdentIsLargerThanPrev;
    if (!isNextLineSetextHeading) {
      break;
    }
  }

  // End list item if found start of new bullet
  if (nextBulletRegex.test(line)) {
    break;
  }

  // ...
}
```

Adding this code solves the issue, and make all the tests pass as well.

I am a bit afraid that the list tokenizer is getting more and more complex, but honestly I don't know if there is any other way (Frankly speaking it seems that markdown syntax is complicated by nature). If there is a better way to do it in marked, I would like to hear about it. Or if this is fine to go on and make a PR?

Thanks.

**Comment by UziTech** [comment]
@jangxyz a PR would be appreciated 😁👍
